Database ingestion and functional API

cxk280 opened the pull request:
This PR handles both database ingestion and the creation of the related API. A few points to keep in mind:

- I have provided a `populateAll` resolver as a convenience to seed all tables. However, seeding `pki_events` can take a long time, so in order to play with this locally more quickly, you can individually run all of the resolvers defined in `src/resolvers/db/index.js` except `populateAll` and `populateDailyCron`. Then search for the comment `// Can use the following line for testing when needed:` in `src/resolvers/db/utils/addToDB.js`. Right below this you can edit the number of rows you want to add to `pki_events`. I find a few hundred to be sufficient to demonstrate the full range of functionality. There will be a console log every 200th iteration so you can keep track of where things are without blowing up your terminal.
- The correct contents of the `eventTypeKey` object in `src/resolvers/db/utils/addToDB.js` is still being sorted out. I emailed Mark about this today and copied you (Josh) on it.
- cron functionality exists and works but is commented out for now so ESLint doesn't yell at me.
- Additional remaining work, such as Jest tests, is tracked in [this Trello board](https://trello.com/b/X5fxVzbe/urbit-network-explorer). I emailed you (Josh) an invite to edit. Let me know if you do not receive that. I try to keep it up to date but I make no assertions that this board reflects all additional work remaining with project as there might be things I'm missing.

Let me know if you have any questions.

galenwp commented:
i'm excited about this

jalehman commented on src/resolvers/api/getNode.js:
```diff
@@ -0,0 +1,147 @@
+
+  const sponsors = []
+
+  // sponsor
+  const getSponsorQuery = `select sponsor_id from pki_events where node_id = '${urbitId}' order by time desc limit 1;`
+  const getSponsorResponse = await query(getSponsorQuery)
+  const getSponsorResponseRow = _get(getSponsorResponse, 'rows[0]') || []
+
+  const sponsor = getSponsorResponseRow ? _get(getSponsorResponseRow, 'sponsor_id') || null : null
+
+  sponsors.push(sponsor)
+
+  // sponsor's sponsor
+  const getSponsorsSponsorQuery = `select sponsor_id from pki_events where node_id = '${sponsor}' order by time desc limit 1;`
+  const getSponsorsSponsorResponse = await query(getSponsorsSponsorQuery)
+  const getSponsorsSponsorResponseRow = _get(getSponsorsSponsorResponse, 'rows[0]') || []
+  const sponsorsSponsor = getSponsorsSponsorResponseRow ? _get(getSponsorsSponsorResponseRow, 'sponsor_id') || null : null
+
+  if (sponsorsSponsor) {
+    sponsors.push(sponsorsSponsor)
+  }
```
A node only has one sponsor. If I was using this API from a client and I wanted to show my sponsor's sponsor, I'd expect to use the API roughly as follows:

```graphql
getNode(input: { urbitId: "~wolref-podlex" }) {
  sponsor {
    urbitId
    sponsor {
      urbitId
    }
  }
}
```

cxk280 replied:
In the [bounty description](https://grants.urbit.org/bounties/1101665172), `get-node` returns `sponsors: [node] (sponsor and sponsor's sponsor, if applicable)`. Does this contradict that? Maybe I'm misunderstanding something.

cxk280 replied:
Since based on discussion in another comment it's OK to leave it as an array of sponsors, can we leave this particular item as-is?

cxk280 replied:
After having spent some time working on the materialized view approach for nodes, I think not having having an array of sponsors is indeed the desirable approach here.

jalehman commented on src/resolvers/api/getNode.js:
```diff
@@ -0,0 +1,147 @@
+        getStatusIdResponse = await query(getStatusIdQuery)
+
+        if (_get(getStatusIdResponse, 'rows[0].count') === '0') {
+          getStatusIdQuery = `select count(*) from pki_events where node_id = '${urbitId}' and address = '0x86cd9cd0992f04231751e3761de45cecea5d1801' or address = '0x8c241098c3d3498fe1261421633fd57986d74aea';`
```
What are these hard-coded addresses?

cxk280 replied:
From an email from Tlon Mark (looks like you were copied):

```
If you want to know whether or not a specific star is *owned* by a lockup contract, you can check to see if its current owner is either of these two addresses:
0x86cd9cd0992f04231751e3761de45cecea5d1801 (linear lockup)
0x8c241098c3d3498fe1261421633fd57986d74aea (conditional lockup)
```

I'll add an explanatory comment in the code.

jalehman commented on src/resolvers/api/getNode.js:
```diff
@@ -0,0 +1,147 @@
+          getStatusIdResponse = await query(getStatusIdQuery)
+
+          if (_get(getStatusIdResponse, 'rows[0].count') === '0') {
+            statusId = 1
+          } else {
+            statusId = 2
+          }
+        } else {
+          statusId = 3
+        }
+      } else {
+        statusId = 4
+      }
+    } else {
+      statusId = 5
```
These are [magic numbers](https://en.wikipedia.org/wiki/Magic_number_(programming)) and should be replaced with constants that refer to them by name.

cxk280 replied:
Good catch! I had these in here as a placeholder at one point and forgot to swap them out for constants. Fixed.

jalehman commented on src/resolvers/api/getNode.js:
```diff
@@ -0,0 +1,147 @@
+    getStatusIdResponse = await query(getStatusIdQuery)
+
+    if (_get(getStatusIdResponse, 'rows[0].count') === '0') {
+      getStatusIdQuery = `select count(*) from pki_events where node_id = '${urbitId}' and event_type_id = '6';`
```
See below comment about magic numbers (referring to `event_Type_id = '6'`).

cxk280 replied:
Fixed.

jalehman commented on src/resolvers/api/getNode.js:
```diff
@@ -0,0 +1,147 @@
+    sponsors.push(sponsorsSponsor)
+  }
+
+  // status
```
This whole status calculation is pretty messy. The nesting is unnecessary and makes things hard to read, as does the rebinding of `getStatusId{Query,Response}`. I'd prefer it if you:

- Broke this out into its own function that, given an `urbitId`, produces the appropriate status.
- Since you're now working with a function, you can `return` a status once you find it.
- If you start with the innermost query (with the hard-coded addresses), you'll know whether or not you need to proceed with subsequent queries. Happy to talk through this during one of our sessions.

cxk280 replied:
Good call. Definitely makes the file look a lot better to abstract that out, and of course any time one avoids nested conditionals is good. I moved the function to `/network-api/src/resolvers/utils.js`. Let me know if you'd prefer it somewhere else.

jalehman commented on src/resolvers/api/getNode.js:
```diff
@@ -0,0 +1,147 @@
+  const getKidsResponseRows = _get(getKidsResponse, 'rows') || []
+  const kids = []
+
+  for (const i in getKidsResponseRows) {
```
See above comment about `for...in` queries on arrays.

cxk280 replied:
Fixed.

jalehman commented on src/resolvers/api/getNode.js:
```diff
@@ -0,0 +1,147 @@
+  const getKidsQuery = `select distinct node_id from pki_events where sponsor_id = '${urbitId}';`
+  const getKidsResponse = await query(getKidsQuery)
+  const getKidsResponseRows = _get(getKidsResponse, 'rows') || []
+  const kids = []
+
+  for (const i in getKidsResponseRows) {
+    const nodeId = _get(getKidsResponseRows, `[${i}].node_id`) || null
+    kids.push(nodeId)
+  }
```
```suggestion
  const kids = getKidsResponseRows.map(row => row.node_id }); 
```

cxk280 replied:
Fixed.

jalehman commented on src/resolvers/api/getNode.js:
```diff
@@ -0,0 +1,147 @@
+  }
+
+  // continuity number
+  const getContinuityNumberQuery = `select continuity_number from pki_events where node_id = '${urbitId}' order by time desc limit 1;`
```
Reading from `pki_events` this way is going to be very inefficient. It means each query of a single node is going to have to result in several database queries, which won't be acceptable with respect to performance. This is why I had a `node` table in the original entity-relationship diagram. There are two approaches that can be used here:

1. After initially populating the `pki_event` table, run another job that populates the `node` table with all of these calculations precomputed. This could be done by parsing the `pki_event` data that originally went into the table, or from querying the table.
2. Create the `node` table as a materialized view of `pki_event`.

I'm generally in favor of (2), as it means we don't have to duplicate any data. It does, however, mean that you'll need to learn how to create one of these. I'd be happy to assist in showing you how to do this.

jalehman replied:
To get an idea of the size of the problem here, this single function issues 15 different database queries. Your implementation of `getNodes` calls this function. This means that a modest query of 10 nodes would result in 150 distinct database queries.

By using either of the strategies mentioned above you could bring this number of queries down to **only one** no matter how many nodes are requested (SQL has built-in limit & offset capabilities).

cxk280 replied:
I definitely want to do option 2 as it seems the best technical approach. I've spent some time trying to cobble the materialized view together and have gotten part of it, but I do think this is something I'll need some support from you on. We can discuss on our upcoming meeting.

cxk280 replied:
I definitely want to do option 2 as it seems the best technical approach. I've spent some time trying to cobble the materialized view together and have gotten part of it, but I do think this is something I'll need some support from you on. We can discuss on our upcoming meeting.

cxk280 replied:
This is now done. Thanks for your help constructing that query!

jalehman commented on index.js:
```diff
@@ -1,23 +1,161 @@
+
+  type Ping {
+    pingId: Int!
+    nodeId: String!
```
This means you're going to need to learn how to use field resolvers (see the code sample under [this header](https://www.apollographql.com/docs/apollo-server/data/resolvers/#example) in the Apollo docs).

cxk280 replied:
Actually, it turns out that the `Ping` type is not needed.

jalehman reviewed (changes requested):
I haven't gotten through the whole thing, but I've already come up with a lot to work on so I'm going to submit the review so far. Many of my comments, e.g. the use of `for...in` loops, will apply to many other areas of your code.

cxk280 commented on index.js:
```diff
@@ -1,23 +1,161 @@
+  }
+  
+  type Node {
+    urbitId: String
```
No. Will change to `String!`.

cxk280 commented on index.js:
```diff
@@ -1,23 +1,161 @@
+  type Node {
+    urbitId: String
+    numOwners: Int
+    sponsors: [String]
```
In the [bounty description](https://grants.urbit.org/bounties/1101665172), `get-node` returns `sponsors: [node] (sponsor and sponsor's sponsor, if applicable)`. In that case, this should be an array. Let me know if I am misunderstanding something.

jalehman replied:
Ah, good catch. I'm okay with changing it to a single `Node` and having the client get the sponsor's sponsor via GraphQL (there's another comment in this review that references this), or keeping it as an array. Regardless though, the type should be a Node and not a String.

cxk280 replied:
If it's all the same, I'd prefer keeping it as an array as this requires me to redo less. I have changed it to `Node` from `String`, though.

cxk280 replied:
After digging deeper into constructing a materialized view for `node`, it is in fact a better approach to use it as a single `Node` instead of an array, so that is what I will do.

cxk280 commented on index.js:
```diff
@@ -1,23 +1,161 @@
+    sponsors: [String]
+    statusId: Int!
+    kids: [String]
+    nodeType: NodeType
```
Fixed.

cxk280 commented on index.js:
```diff
@@ -1,23 +1,161 @@
+
+  type Activity {
+    urbitId: String
+    date: Date
```
Fixed.

cxk280 commented on index.js:
```diff
@@ -1,23 +1,161 @@
+  type Activity {
+    urbitId: String
+    date: Date
+    active: Boolean
```
Fixed.

cxk280 commented on index.js:
```diff
@@ -1,23 +1,161 @@
+const { GraphQLScalarType, Kind } = require('graphql')
+
+// const { startCron } = require('./src/cron')
+const dbResolvers = require('./src/resolvers/db/index')
+const apiResolvers = require('./src/resolvers/api/index')
```
Fixed.

cxk280 commented on src/resolvers/api/getActivity.js:
```diff
@@ -0,0 +1,64 @@
+      }
+    }
+  } else {
+    returnArr.push({ urbitId, active: false, date: null })
```
Fixed.

cxk280 commented:
Please note that I am making additional commits addressing PR feedback just to make it easier to keep track as review is in progress, but happy to squash those when we're ready to merge.

cxk280 commented on index.js:
```diff
@@ -1,23 +1,161 @@
+  
+  type Node {
+    urbitId: String
+    numOwners: Int
```
Good catch. I will make this `Int!`. As for the default, is it possible for a Node to have no owners? If it's spawned, it has to be owned by a wallet, right? If so, shouldn't the default be `1`?

cxk280 commented on index.js:
```diff
@@ -1,23 +1,161 @@
+    # COMET
+  }
+
+  enum EventName {
```
Fixed.

cxk280 commented on index.js:
```diff
@@ -1,23 +1,161 @@
+    revisionNumber: Int
+  }
+
+  type EventType {
+    eventType: Int!
+    eventName: EventName!
+  }
```
If I understand correctly, the consequence of this is that the `EventType` type that this comment references (lines 106-109) should simply be deleted (in addition to the changes requested in other comments, of course). So I will do so. Let me know if this is a misunderstanding.

cxk280 commented on index.js:
```diff
@@ -1,23 +1,161 @@
+  type PKIEvent {
+    eventId: Int!
+    nodeId: String!
+    eventTypeId: Int!
```
Fixed.

cxk280 commented on index.js:
```diff
@@ -1,23 +1,161 @@
+    eventId: Int!
+    nodeId: String!
+    eventTypeId: Int!
+    sponsorId: String
```
Fixed.

cxk280 commented on index.js:
```diff
@@ -1,23 +1,161 @@
+  }
+
+  type Activity {
+    urbitId: String
```
Fixed.

cxk280 commented on src/resolvers/api/getActivity.js:
```diff
@@ -0,0 +1,64 @@
+        response_time = response_time.toISOString().split('T', 1)[0]
+      }
+
+      if (!responseDates.includes(response_time)) {
```
Fixed.

cxk280 commented on src/resolvers/api/getActivity.js:
```diff
@@ -0,0 +1,64 @@
+
+      if (!responseDates.includes(response_time)) {
+        returnArr.push({ urbitId, active: online, date: response_time })
+        responseDates.push(response_time)
```
Fixed. I also changed `response_time` to the camelCase `responseTime` when used as a variable, just to be consistent.

cxk280 commented on src/resolvers/api/getActivity.js:
```diff
@@ -0,0 +1,64 @@
+  const rows = _get(getActivityResponse, 'rows') || []
+
+  if (rows.length > 0) {
+    for (const i in rows) {
```
Happy to change this because that probably is a better way to handle the default value assignment instead of using `_get`. However, `for...in` probably isn't a problem here itself, since the order of elements in the array being iterated over is not important. Right?

cxk280 replied:
I have now changed this for all other occurrences of the `for...in` pattern, except for one case in which I decided `for (let i = 0; i < something.length; i++)` made more sense.

cxk280 commented on src/cron.js:
```diff
@@ -0,0 +1,23 @@
+const cron = require('node-cron')
+
+const cronQuery = `
+  mutation {
+    populateDailyCron
+  }
+`
+
+// Below is every minute for testing
+const cronExpression = '* * * * *'
+
+// Below is every five minutes for testing
+// const cronExpression = '*/5 * * * *'
+
+// Below is every 24 hours at midnight for production
+// const cronExpression = '0 0 * * *'
+
+const startCron = () => {
+  cron.schedule(cronExpression, () => request('http://localhost:4000', cronQuery))
+}
+
+module.exports = { startCron }
+
```
I had the code ready (in a comment) to run it automatically when the server starts, I just had it commented out for your ease when testing. I'll go ahead and change it to always run when the server starts. Good call about just bringing in environment variables to make this more clear. Fixed.

cxk280 commented on src/resolvers/api/getActivity.js:
```diff
@@ -0,0 +1,64 @@
+
+const { query } = require('../utils')
+
+const getActivity = async (_, args) => {
+  const urbitId = _get(args, 'input.urbitId') || null
+  const since = _get(args, 'input.since') || null
+  const until = _get(args, 'input.until') || null
```
Fixed.

cxk280 commented on index.js:
```diff
@@ -1,23 +1,161 @@
+const dbResolvers = require('./src/resolvers/db/index')
+const apiResolvers = require('./src/resolvers/api/index')
 
 const typeDefs = gql`
```
Moving to `src/schema.js` is done.

What form would you like the documentation to take? Simple comments?

cxk280 replied:
I have pushed up schema documentation changes. As these are necessarily somewhat subjective (and such subjectivity seems expected by your comment), I used my best judgement. Feel free to request any changes to them you'd like.

cxk280 commented on index.js:
```diff
@@ -1,23 +1,161 @@
+    populateEventType: Boolean
+    populateNodeType: Boolean
+    populateAll: Boolean
+    populateDailyCron: Boolean
```
Fixed. Removed all mutations and added the relevant scripts in `package.json`.

cxk280 commented on index.js:
```diff
@@ -1,23 +1,161 @@
-  Query: {...dbResolvers, ...apiResolvers}
+  Date: dateScalar,
+  Query: { ...apiResolvers },
+  Mutation: { ...dbResolvers }
```
Fixed.
